File: src/code/js/Q046.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function (nums) {
    var res = [];
    permuteHelper(res, nums, 0);
    return res;
};
var permuteHelper = function (res, nums, pos) {
    if (pos === nums.length - 1) {
        res.push(nums.slice());
    } else {
        for (var i = pos; i < nums.length; i++) {
            var temp = nums[i];
            nums[i] = nums[pos];
            nums[pos] = temp;
            permuteHelper(res, nums, pos + 1);
            nums[pos] = nums[i];
            nums[i] = temp;
        }
    }
};
/*
  执行过程(深度优先)
  pos=0            _________[1,2,3]_________
                  /            |            \
  pos=1     [1,2,3]         [2,1,3]         [3,2,1]
              / \             / \             / \
  pos=2 [1,2,3] [1,3,2] [2,1,3] [2,3,1] [3,2,1] [3,1,2]
 */

File: src/code/java/Q046.java
package code.java;

import java.util.ArrayList;
import java.util.List;

/*
   执行过程(深度优先)
  n=0            _________[1,2,3]_________
                /            |            \
  n=1     [1,2,3]         [2,1,3]         [3,2,1]
            / \             / \             / \
  n=2 [1,2,3] [1,3,2] [2,1,3] [2,3,1] [3,2,1] [3,1,2]
 * */
public class Q046 {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        permute(result, nums, 0);
        return result;
    }

    private void permute(List<List<Integer>> result, int[] nums, int n) {
        if (n == nums.length - 1) {
            List<Integer> list = new ArrayList<>();
            for (int num : nums) {
                list.add(num);
            }
            result.add(list);
            return;
        }
        for (int i = n; i < nums.length; i++) {
            int temp = nums[i];
            nums[i] = nums[n];
            nums[n] = temp;
            permute(result, nums, n + 1);
            nums[n] = nums[i];
            nums[i] = temp;
        }
    }
}
